pkg/ebpf/c/tracee.bpf.c

```diff
@@ -6289,3 +6313,836 @@ int tc_ingress(struct __sk_buff *skb)
+
+    // mark network events to be submitted to userland
+    should_submit_net_event(&neteventctx, data.config);
+
+    // (*) Use skb timestamp as the key for a map shared between this kprobe and
+    // the skb ebpf program: this is **NOT SUPER** BUT, for older kernels, that
+    // provide ABSOLUTE NO eBPF helpers in cgroup/skb programs context, it does
+    // its job: pre-process everything HERE so cgroup/skb programs can use.
+    //
+    // Explanation: The cgroup/skb eBPF program is called right after this
+    //              kprobe, but preemption is enabled. If preemption wasn't
+    //              enabled, we could simply populate a single item map and pick
+    //              pointer inside cgroup/skb. Instead, we index map items
+    //              using the skb timestamp, which is a value that is shared
+    //              among this kprobe AND the cgroup/skb program context
+    //              (through its skbuf copy).
+    //
+    // Theoretically, map collisions might occur, BUT very unlikely due to:
+    //
+    // kprobe (map update) -> cgroup/skb (consume) -> kretprobe (map delete)
+
+    bpf_map_update_elem(&cgrpctxmap, &skbts, &neteventctx, BPF_NOEXIST);
+
+    return 0;
+}
+
+SEC("kretprobe/__cgroup_bpf_run_filter_skb")
+int BPF_KRETPROBE(ret_cgroup_bpf_run_filter_skb)
+{
+    // runs AFTER the CGROUP/SKB eBPF program
+
+    event_data_t data = {};
+    if (!init_event_data(&data, ctx))
+        return 0;
+
+    // Both ingress & egress (mostly ingress) might run from a kthread and still
+    // need processing. Instead of "should_trace" here, check if there is a
+    // current skb timestamp entry (added by the entry) and delete it.
+
+    // pick from entry from entrymap
+    u32 pid = data.context.task.host_pid;
+    struct entry *entry = bpf_map_lookup_elem(&entrymap, &pid);
+    if (!entry) // no entry == no tracing
+        return 0;
+
+    // pick args from entry point's entry
+    // struct sock *sk = (void *) entry->args[0];
+    struct sk_buff *skb = (void *) entry->args[1];
+
+    // cleanup entrymap
+    bpf_map_delete_elem(&entrymap, &pid);
+
+    // use skb timestamp as the key for cgroup/skb
+    u64 skbts = BPF_READ(skb, tstamp);
+
+    // only continue if netctx exists
+    net_event_context_t *netctx = bpf_map_lookup_elem(&cgrpctxmap, &skbts);
+    if (!netctx)
+        return 0;
+
+    // delete netctx after cgroup ebpf program runs
+    bpf_map_delete_elem(&cgrpctxmap, &skbts);
+
+    return 0;
+}
+
+//
+// Type definitions and prototypes for protocol parsing
+//
+
+// NOTE: proto header structs need full type in vmlinux.h (for correct skb copy)
+
+typedef union iphdrs_t {
+    struct iphdr iphdr;
+    struct ipv6hdr ipv6hdr;
+} iphdrs;
+
+typedef union protohdrs_t {
+    struct tcphdr tcphdr;
+    struct udphdr udphdr;
+    struct icmphdr icmphdr;
+    struct icmp6hdr icmp6hdr;
+    union {
+        u8 tcp_extra[40]; // data offset might set it up to 60 bytes
```

Comment:
The union here only exists because other protocols might have "extra" as well, then they will be included in the union (dccp _extra[], let's say). A TCP packet header might be up to 60 bytes long (thus the 40 bytes extra added to the end of the struct).